### ISSUE TITLE
[ansible] Refactor Kubernetes Roles Based on Services

### ISSUE THREAD

**Issue by danehans** [kubernetes/contrib#838]
The current kubernetes ansible roles are organized into masters and nodes. This makes breaking the master/nodes services into individual units of management, deployment, etc.. In the past, I have seen issues with other projects that do not manage services individually. I suggest changing master into individual roles based on master service, such as:
1. kube-apiserver
2. kube-controller-mansger
3. kube-scheduler

The same for the node role:
1. kubelet
2. kube-proxy

and the client role:
1. kubectl

A single master inventory would stay the same.

site manifest (i.e. cluster.yml) would look like:

```
- hosts: all
  gather_facts: false
  sudo: yes
  roles:
    - pre-ansible
  tags:
    - pre-ansible

# Install etcd
- hosts: etcd
  sudo: yes
  roles:
    - etcd
  tags:
    - etcd

- hosts: all
  sudo: yes
  roles:
    - common
    - docker
  tags:
    - docker

# install flannel
- hosts:
    - masters
    - nodes
  sudo: yes
  roles:
    - { role: flannel, when: networking == 'flannel' }
  tags:
    - network-service-install

# install kube master services
- hosts: masters
  sudo: yes
  roles:
    - kube-apiserver
   - kube-controller-manager
   - kube-scheduler
   - kubectl
   - kubernetes-addons
  tags:
    - masters
    - dns

# install kubernetes on the nodes
- hosts: nodes
  sudo: yes
  roles:
    - kubelet
    - kube-proxy
  tags:
    - nodes
```

**Comment by ingvagabund** [comment]
Not sure if you will be able to separate master service without increasing complexity. As both ``master`` and ``node`` roles depend on ``kubernetes`` role, each invocation of ``apiserver``, ``scheduler`` and other playbooks will call ``kubernetes`` and ``common`` role. At the same time, ``kubernetes`` role is responsible for generating of certificates and tokens.

If would be great to make sure the ``kubernetes`` role gets run only once. Besides, the entire ``master`` role gets run one more time when you run ``kubernetes-addons`` playbook. Assuming all tasks are idempotent, it does not matter. Still, you run some tasks even when there is no need for it.

I can think of the following scenarios:

* remove dependency of ``kubernetes`` roles from ``master``/``node`` and run the ``kubernetes`` right before them. E.g.

```yaml
- hosts: masters
  sudo: yes
  roles:
   - kubernetes
   - kube-apiserver
   - kube-controller-manager
   - kube-scheduler
   - kubectl
   - kubernetes-addons
  tags:
```

This is not good, as when you decide to run "just" ``kubernetes-apiserver``, ``kubernetes`` role does not get called.

* create one playbook for each ``kube``-like role as you suggest. The ``kubernetes`` role gets called 7-8 times (or maybe my understanding of the ``meta`` is wrong and it gets called only once). Sometimes, you need the ``kubelet`` installed on the master as well. In a case of containerized kubernetes (which is not currently supported), you need kubelet on the master everytime. Well, in any case, you will need one role for each ``kube*``. So instead of shipping ``4`` roles (``kubernetes``, ``master``, ``node``, ``kubernetes-addons``), we will have ``8``. It is a shame, there is no way to create more entry points for one role (the ``main.yml`` in a role is the only entry point). From the point of reusable, it is complete perfect to have one role for each micro-service. From the point of maintainability, it is a lot of code/task execution duplication. 

* break the playbook into smaller independent playbooks. Instead of running one huge playbook at once, call one smaller playbook inside a script. To deploy the cluster, run a sequence of scripts, each script running a playbook. Here, you will get a lot of repetitive tasks again. The benefit of it is that you can keep all master services inside one role. Depending on the tags you set, only the subset of tasks inside a playbook gets called. This way, you keep the master code together. At the same time, it is transparent for a user. On the other hand, you can not include the smaller playbooks inside of a single playbook.

I would love to see ``kube-scheduler``, ``kube-proxy`` and other roles. However, the roles are so tied together it is impossible to do that without loosing benefits of task/code sharing.

**Comment by ingvagabund** [comment]
Answering myself, the ``kubernetes`` role gets run only once [1]. Unless ``allow_duplicates: yes`` is specified.

[1] http://docs.ansible.com/ansible/playbooks_roles.html#role-dependencies

**Comment by ingvagabund** [comment]
Anyway, https://github.com/kubernetes/contrib/pull/1444 is prerequisite for the ``master`` decomposition. The PR refactors and breaks the ``master`` and ``node`` into more tasks, generalizing the ``main.yml`` file. Once merged, it can be broken out into more roles/micro-services.

**Comment by ingvagabund** [comment]
The ``node``/``master`` can be decomposed into 3 steps:

* tasks common for node/master (collection tokens, certs)
* install, configure and start/restart kube services
* tasks common for node/master (run fluentd, update firewalld and iptables)

The middle step can be made independent. In a case of node, tokens can be collected after kubelet and proxy are running. For the master case, you need apiserver to start in order to generate tokens (prove me wrong). Again, some steps can be postponed. So instead of ``master`` and ``node`` roles that depends on ``kubernetes``, we end up with ``6`` roles, some of them depending on ``kubernetes``, some of them becoming dependence of "some" common role for ``node``, resp. ``master``.

Unless there is a plan to run ``apiserver``, ``scheduler`` and ``controller-manager`` each inside their own VM, we are always going to install all of them on the same node. In the case of containerized kubernetes, we will end up with the same question. I can imagine the case of running ``apiserver`` as a container and ``scheduler`` from installed rpm (but then why not run both of them as containers). Ok, ``kubelet`` is something else. Still, it is a temporary state and in the future, even ``kubelet`` gets containerized. So instead of replacing rpm roles with containerized roles, we can replace entire ``master`` rpm-based role with container-based one.

So, what is the benefit of breaking the ``master`` roles into individual micro-services?
